test/tests/functional/pbs_qstat.py

```diff
@@ -57,6 +58,61 @@ def parse_json(self, dictitems, qstat_attr):
                         self.parse_json(val, qstat_attr)
         return qstat_attr
 
+    def validate_el(self, attr_dict, user):
+        """
+        Check if the value of server attributes reported by
+        qstat -Bf matches the expected values.
+        """
+        qstat_o = {}
+        qstat_o['server_state'] = 'Active'
+        qstat_o['server_host'] = self.server.hostname
+        qstat_o['scheduling'] = 'True'
+        qstat_o['total_jobs'] = '0'
+        qstat_o['state_count'] = ['Transit', 'Queued',
+                                  'Held', 'Waiting', 'Running',
+                                  'Exiting', 'Begun']
+        qstat_o['default_queue'] = 'workq'
+        qstat_o['log_events'] = '511'
+        qstat_o['mail_from'] = 'adm'
+        qstat_o['query_other_jobs'] = 'True'
+        qstat_o['resources_default.ncpus'] = '1'
+        qstat_o['default_chunk.ncpus'] = '1'
+        qstat_o['scheduler_iteration'] = '600'
+        qstat_o['resv_enable'] = 'True'
+        qstat_o['node_fail_requeue'] = '310'
+        qstat_o['max_array_size'] = '10000'
+        qstat_o['pbs_license_linger_time'] = '3600'
+        qstat_o['pbs_version'] = self.server.pbs_version
+        qstat_o['eligible_time_enable'] = 'False'
+        qstat_o['max_concurrent_provision'] = '5'
+
+        if user == ROOT_USER:
+            qstat_o['power_provisioning'] = 'False'
+
+        platform = self.du.get_platform()
+        if platform == 'cray' or platform == 'craysim':
+            qstat_o['restrict_res_to_release_on_suspend'] = 'ncpus'
+
+        for each in attr_dict:
+            if each in qstat_o:
+                if "state_count" in each:
+                    out1 = []
+                    out = attr_dict[each].split()
+                    for k in out:
+                        out1.append(k.split(':')[0])
+                    self.assertListEqual(qstat_o[each], out1,
+                                         "Value of server attribute \"%s\""
+                                         "does not match expected"
+                                         "value" % each)
+                else:
+                    self.assertEqual(qstat_o[each], attr_dict[each],
+                                     "Value of server attribute \"%s\""
+                                     "does not match expected value" % each)
```

Comment:
I think this doesn't quite do the right thing.  This will make sure the correct attributes are in the qstat output, but if an unexpected attribute is there, it will mark that as successfully validated.  I think what you want is to replace the if each in qstat_o with an self.assertIn(each, qstat_o).

Reply:
There are some attributes from qstat -Bf which we are not checking in the test case like flicenses, pbs_license_info. In that case, self.assertIn(each, qstat_o) will throw an exception. To avoid this, we can add all possible attributes in output of "qstat -Bf" to 'qstat_o' and set the value as 'None' for attributes which we don't want to validate.